### PR TITLE
[All] Deployments Table

### PR DESCRIPTION
**Pull request by philipnrmn**
This PR merges the deployments table feature branch into master. Each component PR has already been reviewed. Please do not merge until the following list is complete:

Component PRs:
- [x] #241 [1/8] Deployments Table - Deployment struct updates
- [x] #255 [2/8] Deployments Table - Don't use ServicesList to list services
- [x] #247 [3/8] Deployments Table - Add badge to deployments tab
- [x] #249 [4/8] Deployments Table - Use DCOSStore instead of MarathonStore
- [x] #372 [5] Deployments Table - Introduce service rows
- [x] #378 [6] Deployments Table - improve data interpolation
- [x] #388 [7] Deployments Table - show rollback link
- [x] #401 [8] Deployments Table - styling
- [x] #411 [9] Deployments Table - rollback and remove for 'starting' deployments
- [x] #425 [10] Deployments Table - Replace breadcrumbs with nested services links
- [x] #430 [11] Use store mixin forced updates instead of react lifecycle
- [ ] #??? [12] Handle error response on rollback deployments

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @aldipower

File: src/js/pages/services/DeploymentsTab.js [hunk 1]
```diff
@@ -1,73 +1,158 @@
...
+      {name: 'dcos', events: ['change'], suppressUpdate: true}
+    ];
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    }, this);
+  }
+
+  onDcosStoreChange() {
+    this.setState({deployments: DCOSStore.deploymentsList});
```
Comment (rcorral) [inline src/js/pages/services/DeploymentsTab.js]: We typically avoid this patter of storing data that we already have in components. You could instead remove the `supressUpdate: true` and just get the `deploymentsList` where you need it in this component.
This avoids storing the data in the store and in `n` components that are being rendered.
The component will automatically rerender if you remove the `suppressUpdate`

File: src/js/stores/DCOSStore.js [hunk 2]
```diff
@@ -97,6 +106,37 @@ class DCOSStore extends EventEmitter {
...
+          if (service != null) {
+            memo.push(service);
+          }
+          return memo;
+        }, []);
+
+        return Object.assign({affectedServices: services}, deployment);
+      });
+
+    this.emit(DCOS_CHANGE);
```
Comment (rcorral) [inline src/js/stores/DCOSStore.js]: Would it be better to have a different event name? Otherwise anything that listens for `DCOS_CHANGE`.
We typically have a new event name per endpoint and it looks like the deployments data comes from a different endpoint than the list of running apps.

File: src/js/pages/services/DeploymentsTab.js [hunk 3]
```diff
@@ -1,73 +1,158 @@
...
+  }
+
+  renderAffectedServices(prop, deployment) {
+    return (
+      <dl className="deployment-services-list flush-top flush-bottom tree-list">
+        <dt className="deployment-id text-uppercase">{deployment.id}</dt>
```
Comment (rcorral) [inline src/js/pages/services/DeploymentsTab.js]: Can you use the `#getId` method on the Deployment struct?

File: src/js/pages/services/DeploymentsTab.js [hunk 4]
```diff
@@ -1,73 +1,158 @@
...
+      );
+    });
+  }
+
+  renderLocation(prop, deployment) {
+    const services = deployment.getAffectedServices();
+    const items = services.map(function (service, index) {
+      return (
+        <li key={index}>
+          <ServicesBreadcrumb
```
Comment (rcorral) [inline src/js/pages/services/DeploymentsTab.js]: We should be using the new `Breadcrumbs` component, all of the new components will be using it and we will be removing the old `ServicesBreadcrumb`. See this #416

File: src/js/stores/DCOSStore.js (hunk 2 above)
Comment (pierlo-upitup) [inline src/js/stores/DCOSStore.js]: We are doing this as part of the `DCOS` omnistore and this is consistent with what's in the rest of the file:

https://github.com/dcos/dcos-ui/blob/master/src/js/stores/DCOSStore.js#L108
https://github.com/dcos/dcos-ui/blob/master/src/js/stores/DCOSStore.js#L122
https://github.com/dcos/dcos-ui/blob/master/src/js/stores/DCOSStore.js#L163
etc...

File: src/js/pages/services/DeploymentsTab.js (hunk 4 above)
Comment (pierlo-upitup) [inline src/js/pages/services/DeploymentsTab.js]: Let's do that in a separate PR.

Comment (philipnrmn) [inline src/js/pages/services/DeploymentsTab.js]: #425

**Comment by philipnrmn** [comment]
NB tests will fail until #425 is merged.

File: src/js/stores/DCOSStore.js (hunk 2 above)
Comment (philipnrmn) [inline src/js/stores/DCOSStore.js]: Yeah I was aiming for consistency with the Marathon Groups data which is used in the Services tab. I would like to see more fine-grained events eventually though.

File: src/js/pages/services/DeploymentsTab.js (hunk 3 above)
Comment (philipnrmn) [inline src/js/pages/services/DeploymentsTab.js]: Addressed directly on this branch in fbc6a0e

File: src/js/pages/services/DeploymentsTab.js (hunk 1 above)
Comment (philipnrmn) [inline src/js/pages/services/DeploymentsTab.js]: Happy to respect the convention if it's genuinely preferred, but I'd like to understand what the benefit is of breaking with React idiom. Forcing updates via the store mixin makes it hard for me to understand the update/render cycle and circumvents `shouldComponentUpdate`.

Comment (rcorral) [inline src/js/pages/services/DeploymentsTab.js]: Yes, please let's keep things consistent. At the moment it is the preferred way of doing things.
A few thoughts:
* Storing the same datum in multiple components isn't a good practice
* If we truly don't want to rerender when an event gets fired through Storemixin then let's use `suppressUpdate` but it looks like we do want this behavior right now.
* ForceUpdate isn't pretty, I agree. Let's consider other solutions for the future.

File: src/js/stores/DCOSStore.js (hunk 2 above)
Comment (rcorral) [inline src/js/stores/DCOSStore.js]: Fair, it just sucks that so much stuff renders all the time :(

File: src/js/pages/services/DeploymentsTab.js (hunk 1 above)
Comment (philipnrmn) [inline src/js/pages/services/DeploymentsTab.js]: #430

**Comment by pierlo-upitup** [comment]
Seems to me all comments have been addressed minus `Handle error response on rollback deployments`. Can we merge this and address that outside of this PR please?

**Comment by aldipower** [comment]
Very nice work! :ok_hand:

**Comment by aldipower** [comment]
retest this please
